src/programs/bamsormadup.1
.TH BAMSORMADUP 1 "April 2015" BIOBAMBAM
.SH NAME
bamsormadup - sort name collated SAM or BAM file by coordinate and mark duplicates or sort SAM or BAM file by query name
.SH SYNOPSIS
.PP
.B bamsormadup [options]
.SH DESCRIPTION
bamsormadup has two modes of operation depending on the value of the SO
parameter. If SO=coordinate or if the SO key is not given, then
bamsormadup reads a name collated BAM or SAM file from standard input, runs a fix mate
process, sorts the contained alignments by coordinate, marks duplicate
alignments and writes the sorted alignments to standard output in BAM
format. An alignment file is name collated if all the alignments for one
read name appear consecutively in the file. If SO=queryname then the program
reads a BAM or SAM file from standard input, sorts it by queryname and writes the
sorted file on standard output in BAM format.
.PP
The following key=value pairs can be given:
.PP
.B level=<-1|0|1|9|11>:
set compression level of the output BAM file. Valid
values are
.IP -1:
zlib/gzip default compression level
.IP 0:
uncompressed
.IP 1:
zlib/gzip level 1 (fast) compression
.IP 9:
zlib/gzip level 9 (best) compression
.P
If libmaus has been compiled with support for igzip (see
https://software.intel.com/en-us/articles/igzip-a-high-performance-deflate-compressor-with-optimizations-for-genomic-data)
then an additional valid value is
.IP 11:
igzip compression
.PP
.B inputformat=<bam>: 
set the input file format.
This can be either bam or sam (see http://samtools.sourceforge.net/SAM1.pdf)
.PP
.B threads=<[1]>:
number of threads used
.PP
.B M=<stderr>: 
name of the metrics file for duplicate marking (metrics are written to standard error if not set)
.PP
.B tmpfile=<bamsormadup_hostname_pid_starttime>:
prefix for temporary files. By default the temporary files are created in the current directory.
Set 
.nh
tmpfile=mem:tmp_ 
.hy
to store temporary files in RAM instead of on disk. Note that this may require very large amounts of RAM depending on the input.
.PP
.B SO=<coordinate|queryname>: 
set the sort order. Valid values are
.IP coordinate
sort alignments by coordinate. Input is assumed to be name collated.
.IP queryname
sort alignments by query name. No assumption is made on the order of the input.
.PP
.B reference=<>: 
name of reference FastA file when writing CRAM. This file will be used for
filling missing UR and M5 fields of SQ header lines. It may refer to a local
file or a file stored on an http or ftp server. The file is uncompressed on
the fly if the file name ends on .gz . If the REF_CACHE environment variable
is set to the name of an existing directory, then normalised cache files
will be written to this directory for each reference sequence. The file
names are constructed from the directory name and the MD5 checksum of each
reference sequence. This writing of cached files is omitted however, if a
previously existing file is found in the list of read only cache locations
given by the REF_PATH environment variable.
.PP
.B optminpixeldif=<100>:
distance (x and y inside same tile) inside which reads are considered as
optical duplicates
.PP
.B rcsupport==<0>:
if 1 then create rc aux field (unclipped coordinate) for mapped reads when
sorting from query to coordinate order
.SH AUTHOR
Written by German Tischler.
.SH "REPORTING BUGS"
Report bugs to <tischler@mpi-cbg.de>
.SH COPYRIGHT
Copyright \(co 2009-2015 German Tischler, \(co 2011-2015 Genome Research Limited.
License GPLv3+: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


debian/changelog
biobambam2 (2.0.76) unstable; urgency=medium

  * add rcsupport switch in bamsormadup

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 08 Aug 2017 16:49:12 +0200

biobambam2 (2.0.75) unstable; urgency=medium

  * add mc field in bammarkduplicatesopt

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 03 Aug 2017 12:24:54 +0200

biobambam2 (2.0.74) unstable; urgency=medium

  * add bamfillquery

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 02 Aug 2017 10:00:47 +0200

biobambam2 (2.0.73) unstable; urgency=medium

  * add bamfilterrefid
  * allow O key in bammerge (disable check for stdout being a terminal if O key is set)
  * add bamrefinterval program

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 20 Jun 2017 14:48:51 +0200

biobambam2 (2.0.72) unstable; urgency=medium

  * add output file prefix for default read group in bamtofastq

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 16 Mar 2017 10:22:23 +0100

biobambam2 (2.0.71) unstable; urgency=medium

  * add outputperreadgrouprgsm and outputperreadgroupprefix options in bamtofastq

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 16 Mar 2017 10:06:40 +0100

biobambam2 (2.0.70) unstable; urgency=medium

  * add bamscrapcount

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 07 Mar 2017 17:05:10 +0100

biobambam2 (2.0.69) unstable; urgency=medium

  * encode ref id of first read in file in bamexplode file names

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 27 Jan 2017 13:32:37 +0100

biobambam2 (2.0.68) unstable; urgency=medium

  * added HD line header merging in bamcat

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 27 Jan 2017 11:33:22 +0100

biobambam2 (2.0.67) unstable; urgency=medium

  * no adapter clipping if remaining sequence would be a single base (fix by Andrew Whitwham)
  * add bamauxmerge

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 26 Jan 2017 17:30:44 +0100

biobambam2 (2.0.66) unstable; urgency=medium

  * add parserbuffersize parameter in bamsormadup 
  * add optminpixeldif parameter passing in bamsormadup
  * add bamfilterlength program

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 05 Jan 2017 12:06:53 +0100

biobambam2 (2.0.65) unstable; urgency=medium

  * fix bug in bammarkduplicatesopt where last set of duplicates was no marked with od fields

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 30 Nov 2016 12:14:14 +0100

biobambam2 (2.0.64) unstable; urgency=medium

  * add addmatecigar option in bammarkduplicatesopt

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 29 Nov 2016 12:31:07 +0100

biobambam2 (2.0.63) unstable; urgency=medium

  * fix bug in bammarkduplicates/bammarkduplicates2 concerning input via pipe, markthreads>1 and rewritebam=2 

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 28 Nov 2016 10:47:09 +0100

biobambam2 (2.0.62) unstable; urgency=medium

  * fix bin computation for unmapped reads in bamsormadup with SAM input 

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 25 Nov 2016 13:34:04 +0100

biobambam2 (2.0.61) unstable; urgency=medium

  * filter out broken alignments (cigar reaching past end of reference) in bamrecalculatecigar instead of quitting

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 21 Nov 2016 09:06:09 +0100

biobambam2 (2.0.60) unstable; urgency=medium

  * add bamnumericalindex program

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 17 Nov 2016 15:13:15 +0100

biobambam2 (2.0.59) unstable; urgency=medium

  * adapt bamparsort to new libmaus2 API
  * add numeric index construction in bamindex

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 14 Nov 2016 14:09:29 +0100

biobambam2 (2.0.58) unstable; urgency=medium

  * Update to libmaus2 version 2.0.268 (fixes for mate cigar string copying in bamsormadup)

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 01 Nov 2016 10:35:57 +0100

biobambam2 (2.0.57) unstable; urgency=medium

  * Added an explicit iRODS disconnect to avoid the iRODS 4.19 crash. (patch by Andrew Whitwham)

 -- German Tischler <tischler@mpi-cbg.de>  Sun, 18 Sep 2016 18:56:06 +0200

biobambam2 (2.0.56) unstable; urgency=medium

  * fix compilation after API change in libmaus2

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 25 Aug 2016 15:51:00 +0200

biobambam2 (2.0.55) unstable; urgency=medium

  * conditional compilation for PTHREAD_MUTEX_RECURSIVE/PTHREAD_MUTEX_RECURSIVE_NP

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 03 Aug 2016 12:12:34 +0200

biobambam2 (2.0.54) unstable; urgency=medium

  * implement the tags option in bamtofastq (support copying of aux fields from SAM/BAM/CRAM to FastQ)

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 02 Aug 2016 16:25:48 +0200

biobambam2 (2.0.53) unstable; urgency=medium

  * update to libmaus2 version 2.0.240
  * merge/import Andrew Whitwham's parallel adapter finding code

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 01 Aug 2016 10:41:37 +0200

biobambam2 (2.0.52) unstable; urgency=medium

  * add bamfiltermc program

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 15 Jul 2016 12:39:39 +0200

biobambam2 (2.0.51) unstable; urgency=medium

  * only copy mate cigar string if mate is mapped in bamsort

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 15 Jul 2016 10:43:26 +0200

biobambam2 (2.0.50) unstable; urgency=medium

  * do not copy CIGAR data of unmapped mates in bamsormadup 

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 05 Jul 2016 16:13:07 +0200

biobambam2 (2.0.49) unstable; urgency=medium

  * update to libmaus2 version 2.0.216 

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 29 Jun 2016 10:38:45 +0200

biobambam2 (2.0.48) unstable; urgency=medium

  * use AdapterFilter object as const in bamadapterfind
  * add blocksortverbose parameter in bamsormadup
  * reflect API update in libmaus2

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 28 Jun 2016 12:28:37 +0200

biobambam2 (2.0.47) unstable; urgency=medium


src/programs/bamsormadup.cpp
/**
    bambam
    Copyright (C) 2009-2015 German Tischler
    Copyright (C) 2011-2015 Genome Research Limited

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
**/
#include "config.h"

#include <libmaus2/bambam/BamBlockWriterBaseFactory.hpp>

#include <libmaus2/bambam/parallel/BlockSortControl.hpp>
#include <libmaus2/bambam/parallel/BlockMergeControl.hpp>
#include <libmaus2/bambam/parallel/FragmentAlignmentBufferPosComparator.hpp>
#include <libmaus2/bambam/parallel/FragmentAlignmentBufferQueryNameComparator.hpp>

#include <libmaus2/digest/Digests.hpp>

#if defined(LIBMAUS2_HAVE_SHA2_ASSEMBLY)
#include <libmaus2/digest/DigestFactory_SHA2_ASM.hpp>
#endif

#if defined(LIBMAUS2_HAVE_SMMINTRIN_H) && defined(HAVE_SSE4)
#include <libmaus2/digest/DigestFactory_CRC32C_SSE42.hpp>
#endif

#include <libmaus2/digest/DigestFactoryContainer.hpp>

#include <libmaus2/parallel/NumCpus.hpp>

#include <libmaus2/util/ArgInfo.hpp>
#include <biobambam2/BamBamConfig.hpp>
#include <biobambam2/Licensing.hpp>

static int getDefaultLevel() { return Z_DEFAULT_COMPRESSION; }
static int getDefaultTempLevel() { return Z_BEST_SPEED; }
// static std::string getDefaultSortOrder() { return "coordinate"; }
static std::string getDefaultInputFormat() { return "bam"; }
static std::string getDefaultSeqChksumHash() { return "crc32prod"; }
static std::string getDefaultDigest() { return "md5"; }
static int getDefaultBlockSortVerbose() { return 0; }
static int getDefaultOptMinPixelDif() { return 100; }
static int getDefaultRCSupport() { return 0; }

template<typename heap_element_type>
static int
	mergeBlocks
(
	libmaus2::parallel::SimpleThreadPool & STP,
	std::ostream & out,
	libmaus2::autoarray::AutoArray<char> & sheader,
	std::vector<libmaus2::bambam::parallel::GenericInputControlStreamInfo> const & BI,
	libmaus2::bitio::BitVector::unique_ptr_type const & Pdupvec,
	int level,
	uint64_t inputblocksize,
	uint64_t inputblocksperfile,
	uint64_t mergebuffersize,
	uint64_t mergebuffers,
	uint64_t complistsize,
	std::string const & hash,
	std::string const & headerchecksumstr,
	std::string const & digesttype,
	std::string const & indextmpfileprefix,
	std::string const & indexfilename,
	std::string const & digestfilename,
	libmaus2::bambam::parallel::BlockMergeControlTypeBase::block_merge_output_format_t const oformat,
	bool const bamindex,
	std::string const & sortordername
)
{
	typedef libmaus2::digest::DigestInterface digest_interface_type;
	typedef digest_interface_type::unique_ptr_type digest_interface_pointer_type;
	digest_interface_pointer_type Pdigest(::libmaus2::digest::DigestFactoryContainer::construct(digesttype));

	bool const computerefidintervals = (oformat == libmaus2::bambam::parallel::BlockMergeControlTypeBase::output_format_cram) && (sortordername == "coordinate");

	libmaus2::bambam::parallel::BlockMergeControl<heap_element_type> BMC(
		STP, // libmaus2::parallel::SimpleThreadPool &
		out, // std::ostream &
		sheader, // libmaus2::autoarray::AutoArray<char> const &
		BI, // std::vector<libmaus2::bambam::parallel::GenericInputControlStreamInfo> const &
		Pdupvec.get(), // libmaus2::bitio::BitVector::unique_ptr_type const &
		level, // int
		inputblocksize, // uint64_t
		inputblocksperfile /* uint64_t, blocks per channel */,
		mergebuffersize /* uint64_t, merge buffer size */,
		mergebuffers /* uint64_t, number of merge buffers */,
		complistsize /* uint64_t, number of bgzf preload blocks */,
		hash, // std::string
		indextmpfileprefix,
		Pdigest.get(),
		oformat,
		bamindex,
		computerefidintervals
	);
	BMC.addPending();
	BMC.waitWritingFinished();
	std::ostringstream finalheaderchecksumstr;
	BMC.printChecksumsForBamHeader(finalheaderchecksumstr);
	// std::cerr << finalheaderchecksumstr.str();
	if ( indexfilename.size() )
		BMC.writeBamIndex(indexfilename);
	std::string const digeststr = BMC.getFileDigest();
	if ( digestfilename.size() )
	{
		libmaus2::aio::OutputStreamInstance PFOS(digestfilename);
		PFOS << digeststr << std::endl;
		PFOS.flush();
	}

	std::cerr << "[D]\t" << digesttype << "\t" << digeststr << std::endl;

	if ( finalheaderchecksumstr.str() != headerchecksumstr )
	{
		std::cerr << "[W] checksum mismatch between input and output" << std::endl;
		return EXIT_FAILURE;
	}
	else
	{
		std::cerr << "[V] checksum ok" << std::endl;
		return EXIT_SUCCESS;
	}
}

template<typename order_type, typename heap_element_type, bool create_dup_mark_info>
int bamsormadupTemplate(
	::libmaus2::util::ArgInfo const & arginfo,
	bool const bamindex,
	bool const fixmates,
	bool const dupmarksupport,
	unsigned int const optminpixeldif,
	std::string const & sortordername,
	bool const rcsupport
)
{
	int returncode = EXIT_SUCCESS;

	libmaus2::timing::RealTimeClock progrtc; progrtc.start();
	// typedef libmaus2::bambam::parallel::FragmentAlignmentBufferPosComparator order_type;
	// typedef libmaus2::bambam::parallel::FragmentAlignmentBufferNameComparator order_type;

	libmaus2::timing::RealTimeClock rtc;

	rtc.start();
	uint64_t const numlogcpus = arginfo.getValue<int>("threads",libmaus2::parallel::NumCpus::getNumLogicalProcessors());
	libmaus2::aio::PosixFdInputStream in(STDIN_FILENO,256*1024);
	std::string const tmpfilebase = arginfo.getUnparsedValue("tmpfile",arginfo.getDefaultTmpFileName());
	int const templevel = arginfo.getValue<int>("templevel",getDefaultTempLevel());
	std::string const seqchksumhash = arginfo.getValue<std::string>("seqchksumhash",getDefaultSeqChksumHash());
	std::string const digest = arginfo.getValue<std::string>("digest",getDefaultDigest());
	std::string const indexfilename = arginfo.getUnparsedValue("indexfilename",std::string());
	std::string const digestfilename = arginfo.getUnparsedValue("digestfilename",std::string());

	std::string const sinputformat = arginfo.getUnparsedValue("inputformat",getDefaultInputFormat());
	libmaus2::bambam::parallel::BlockSortControlBase::block_sort_control_input_enum inform = libmaus2::bambam::parallel::BlockSortControlBase::block_sort_control_input_bam;
	int const blocksortverbose = arginfo.getValue<int>("blocksortverbose",getDefaultBlockSortVerbose());
	uint64_t const parsebuffersize = arginfo.getValueUnsignedNumeric<uint64_t>("parsebuffersize",libmaus2::bambam::parallel::BlockSortControl<order_type,create_dup_mark_info>::getParseBufferSize());

	if ( sinputformat == "bam" )
	{
		inform = libmaus2::bambam::parallel::BlockSortControlBase::block_sort_control_input_bam;
	}
	else if ( sinputformat == "sam" )
	{
		inform = libmaus2::bambam::parallel::BlockSortControlBase::block_sort_control_input_sam;
	}
	else
	{
		libmaus2::exception::LibMausException lme;
		lme.getStream() << "Unknown input format " << sinputformat << std::endl;
		lme.finish();
		throw lme;
	}

	libmaus2::parallel::SimpleThreadPool STP(numlogcpus);
	typename libmaus2::bambam::parallel::BlockSortControl<order_type,create_dup_mark_info>::unique_ptr_type VC(
		new libmaus2::bambam::parallel::BlockSortControl<order_type,create_dup_mark_info>(
			inform,STP,in,templevel,tmpfilebase,seqchksumhash,
			fixmates,
			dupmarksupport,
			optminpixeldif,
			parsebuffersize,
			rcsupport
		)
	);
	VC->setVerbose(blocksortverbose);
	VC->enqueReadPackage();
	VC->waitDecodingFinished();
	// VC->printChecksums(std::cerr);
	std::ostringstream headerchecksumstr;
	VC->printChecksumsForBamHeader(headerchecksumstr);
	// std::cerr << headerchecksumstr.str();
	// VC->printSizes(std::cerr);
	// VC->printPackageFreeListSizes(std::cerr);
	#if defined(AUTOARRAY_TRACE)
	libmaus2::autoarray::autoArrayPrintTraces(std::cerr);
	#endif
	VC->freeBuffers();

	/**
	 * set up metrics stream
	 **/
	::libmaus2::aio::OutputStreamInstance::unique_ptr_type pM;
	std::ostream * pmetricstr = 0;

	if ( arginfo.hasArg("M") && (arginfo.getValue<std::string>("M","") != "") )
	{
		::libmaus2::aio::OutputStreamInstance::unique_ptr_type tpM(
                                new ::libmaus2::aio::OutputStreamInstance(arginfo.getValue<std::string>("M",std::string("M")))
                        );
		pM = UNIQUE_PTR_MOVE(tpM);
		pmetricstr = pM.get();
	}
	else
	{
		pmetricstr = & std::cerr;
	}

	std::ostream & metricsstr = *pmetricstr;

	if ( create_dup_mark_info )
	{
		VC->flushReadEndsLists();
		VC->mergeReadEndsLists(metricsstr,"bamsormadup");
	}

	metricsstr.flush();
	pM.reset();

	std::vector<libmaus2::bambam::parallel::GenericInputControlStreamInfo> const BI = VC->getBlockInfo();
	libmaus2::bitio::BitVector::unique_ptr_type Pdupvec(VC->releaseDupBitVector());
	libmaus2::bambam::BamHeader::unique_ptr_type Pheader(VC->getHeader());
	::libmaus2::bambam::BamHeader::unique_ptr_type uphead(libmaus2::bambam::BamHeaderUpdate::updateHeader(arginfo,*Pheader,"bamsormadup",PACKAGE_VERSION));
	uphead->changeSortOrder(sortordername /* "coordinate" */);
	uphead->text = uphead->filterOutChecksum(uphead->text);
	uphead->text += headerchecksumstr.str();

	libmaus2::bambam::parallel::BlockMergeControlTypeBase::block_merge_output_format_t oformat = libmaus2::bambam::parallel::BlockMergeControlTypeBase::output_format_bam;

	if ( arginfo.getUnparsedValue("outputformat","bam") == "sam" )
		oformat = libmaus2::bambam::parallel::BlockMergeControlTypeBase::output_format_sam;
	if ( arginfo.getUnparsedValue("outputformat","bam") == "cram" )
		oformat = libmaus2::bambam::parallel::BlockMergeControlTypeBase::output_format_cram;

	std::string const reference = arginfo.getUnparsedValue("reference",std::string());
	if ( oformat == libmaus2::bambam::parallel::BlockMergeControlTypeBase::output_format_cram )
	{
		try
		{
			uphead->checkSequenceChecksums(reference);

			if ( ! uphead->checkSequenceChecksumsCached(false /* throw */) )
			{
				char const * refcache = getenv("REF_CACHE");

				if ( (! refcache) || (!*refcache) )
				{
					libmaus2::exception::LibMausException lme;
					lme.getStream() << "Sequence cache is missing sequences but REF_CACHE is not set" << std::endl;
					lme.finish();
					throw lme;
				}

				// try to fill cache
				uphead->getSequenceURSet(true);
			}

			if ( ! uphead->checkSequenceChecksumsCached(true /* throw */) )
			{
				libmaus2::exception::LibMausException lme;
				lme.getStream() << "Sequence cache is missing sequences" << std::endl;
				lme.finish();
				throw lme;
			}
		}
		catch(...)
		{
			STP.terminate();
			STP.join();
			throw;
		}
	}

	std::ostringstream hostr;
	uphead->serialise(hostr);
	std::string const hostrstr = hostr.str();
	libmaus2::autoarray::AutoArray<char> sheader(hostrstr.size(),false);
	std::copy(hostrstr.begin(),hostrstr.end(),sheader.begin());
	VC.reset();

	std::cerr << "[V] blocks generated in time " << rtc.formatTime(rtc.getElapsedSeconds()) << std::endl;

	rtc.start();
	uint64_t const mergeinputmaxmem = arginfo.getValueUnsignedNumeric("mergeinputmaxmem", 8ull*1024ull * 1024ull * 1024ull);
	uint64_t inputblocksize = 1024*1024;
	uint64_t inputblocksperfile = 8;

	while (
		BI.size() * inputblocksize * inputblocksperfile > mergeinputmaxmem
		&&
		inputblocksperfile > 2
	)
	{
		inputblocksperfile -= 1;
	}

	while (
		BI.size() * inputblocksize * inputblocksperfile > mergeinputmaxmem
		&&
		inputblocksize >= 256ull*1024ull
	)
		inputblocksize /= 2;

	while ( BI.size() * inputblocksize * (inputblocksperfile+1) <= mergeinputmaxmem )
		inputblocksperfile += 1;

	std::cerr << "[V] number of blocks to be merged is " << BI.size() << " using " << inputblocksperfile << " blocks per input with block size " << inputblocksize << std::endl;

	if ( BI.size() * inputblocksize * inputblocksperfile > mergeinputmaxmem )
		std::cerr << "[V] warning: " << mergeinputmaxmem << " bytes insufficient for merging" << std::endl;

	uint64_t const mergebuffermemory = arginfo.getValueUnsignedNumeric("mergebuffermemory", 1024ull * 1024ull * 1024ull);
	uint64_t const complistsize = 32;
	int const level = arginfo.getValue<int>("level",Z_DEFAULT_COMPRESSION);

	uint64_t const mergebuffers =
		(oformat == libmaus2::bambam::parallel::BlockMergeControlTypeBase::output_format_cram)
		?
		(2*STP.getNumThreads())
		:
		4
	;
	uint64_t const mergebuffersize = std::max(mergebuffermemory / mergebuffers,static_cast<uint64_t>(1ull));

	try
	{
		mergeBlocks<heap_element_type>(
			STP,std::cout,sheader,BI,Pdupvec,level,inputblocksize,inputblocksperfile,mergebuffersize,mergebuffers,complistsize,seqchksumhash,headerchecksumstr.str(),
			digest,
			tmpfilebase+"_index",
			indexfilename,
			digestfilename,
			oformat,
			bamindex,
			sortordername
		);
	}
	catch(...)
	{
		STP.terminate();
		STP.join();
		throw;
	}

	std::cerr << "[V] blocks merged in time " << rtc.formatTime(rtc.getElapsedSeconds()) << std::endl;

	STP.terminate();
	STP.join();

	std::cerr << "[V] run time " << progrtc.formatTime(progrtc.getElapsedSeconds()) << " (" << progrtc.getElapsedSeconds() << " s)" << "\t" << libmaus2::util::MemUsage() << std::endl;

	return returncode;
}

int bamsormadup(::libmaus2::util::ArgInfo const & arginfo)
{
	std::string const so = arginfo.getUnparsedValue("SO","coordinate");
	unsigned int const optminpixeldif = arginfo.getValueUnsignedNumeric<uint64_t>("optminpixeldif",getDefaultOptMinPixelDif());
	int const rcsupport = arginfo.getValueUnsignedNumeric<uint64_t>("rcsupport",getDefaultRCSupport());

	if ( so == "queryname" )
		return bamsormadupTemplate<
				libmaus2::bambam::parallel::FragmentAlignmentBufferQueryNameComparator,
				libmaus2::bambam::parallel::GenericInputControlMergeHeapEntryQueryName,
				false /* create dup mark info */>(arginfo,false /* bam index */,false /* fix mates */,false /* dup mark support */,0 /* optminpixeldif */,"queryname",false);
	else
		return bamsormadupTemplate<
				libmaus2::bambam::parallel::FragmentAlignmentBufferPosComparator,
				libmaus2::bambam::parallel::GenericInputControlMergeHeapEntryCoordinate,
				true /* create dup mark info */>(arginfo,true /* bam index */,true /* fix mates */,true /* dup mark support */,optminpixeldif,"coordinate",rcsupport);
}

int main(int argc, char * argv[])
{
	try
	{
		#if defined(LIBMAUS2_HAVE_SHA2_ASSEMBLY)
		libmaus2::digest::DigestFactoryContainer::addFactories(libmaus2::digest::DigestFactory_SHA2_ASM());
		#endif
		#if defined(LIBMAUS2_HAVE_SMMINTRIN_H) && defined(HAVE_SSE4)
		libmaus2::digest::DigestFactoryContainer::addFactories(libmaus2::digest::DigestFactory_CRC32C_SSE42());
		#endif

		::libmaus2::util::ArgInfo const arginfo(argc,argv);

		for ( uint64_t i = 0; i < arginfo.restargs.size(); ++i )
			if (
				arginfo.restargs[i] == "-v"
				||
				arginfo.restargs[i] == "--version"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				return EXIT_SUCCESS;
			}
			else if (
				arginfo.restargs[i] == "-h"
				||
				arginfo.restargs[i] == "--help"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				std::cerr << std::endl;
				std::cerr << "Key=Value pairs:" << std::endl;
				std::cerr << std::endl;

				std::vector< std::pair<std::string,std::string> > V;

				V.push_back ( std::pair<std::string,std::string> ( "level=<["+::biobambam2::Licensing::formatNumber(getDefaultLevel())+"]>", libmaus2::bambam::BamBlockWriterBaseFactory::getBamOutputLevelHelpText() ) );
				V.push_back ( std::pair<std::string,std::string> ( "templevel=<["+::biobambam2::Licensing::formatNumber(getDefaultTempLevel())+"]>", "compression setting for temporary files (see level for options)" ) );

				V.push_back ( std::pair<std::string,std::string> ( "threads=<["+::biobambam2::Licensing::formatNumber(libmaus2::parallel::NumCpus::getNumLogicalProcessors())+"]>", "number of threads" ) );

				// V.push_back ( std::pair<std::string,std::string> ( "SO=<["+getDefaultSortOrder()+"]>", "sorting order (coordinate or queryname)" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("tmpfile=<[")+arginfo.getDefaultTmpFileName()+"]>", "prefix for temporary files, default: create files in current directory" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("inputformat=<[")+getDefaultInputFormat()+"]>", std::string("input format (sam,bam)") ) );
				V.push_back ( std::pair<std::string,std::string> ( "M=<filename>", "metrics file, stderr if unset" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("seqchksumhash=<[")+getDefaultSeqChksumHash()+"]>", "seqchksum digest function: " + libmaus2::bambam::ChecksumsFactory::getSupportedHashVariantsList()) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("digest=<[")+getDefaultDigest()+"]>", "hash digest computed for output stream (md5, sha512)") );
				V.push_back ( std::pair<std::string,std::string> ( std::string("digestfilename=<[]>"), "name of file for storing hash digest computed for output stream (not stored by default)" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("indexfilename=<[]>"), "name of file for storing BAM index (not stored by default)" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("SO=<coordinate|queryname>"), "output sort order (coordinate by default)" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("outputformat=<[bam]>"), std::string("output format (sam,bam,cram)") ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("reference=<[]>"), std::string("reference FastA for writing cram") ) );
				V.push_back ( std::pair<std::string,std::string> ( "optminpixeldif=<["+::biobambam2::Licensing::formatNumber(getDefaultOptMinPixelDif())+"]>", "pixel difference threshold for optical duplicates (default: 100)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "rcsupport=<["+::biobambam2::Licensing::formatNumber(getDefaultRCSupport())+"]>", "add rc tag (unclipped coordinate, default: 0)" ) );

				::biobambam2::Licensing::printMap(std::cerr,V);

				std::cerr << std::endl;
				return EXIT_SUCCESS;
			}

		return bamsormadup(arginfo);
	}
	catch(std::exception const & ex)
	{
		std::cerr << ex.what() << std::endl;
		return EXIT_FAILURE;
	}
}
